core/src/com/unciv/logic/BarbarianManager.kt

```diff
@@ -0,0 +1,265 @@
+import com.unciv.models.metadata.GameSpeed
+import java.util.*
+import kotlin.math.min
+import kotlin.math.pow
+
+class BarbarianManager {
+    val camps = ArrayList<Encampment>()
+
+    @Transient
+    lateinit var gameInfo: GameInfo
+
+    @Transient
+    lateinit var tileMap: TileMap
+
+    fun clone(): BarbarianManager {
+        val toReturn = BarbarianManager()
+        toReturn.camps.addAll(camps.map { it.clone() } )
+        return toReturn
+    }
+
+    fun setTransients(gameInfo: GameInfo) {
+        this.gameInfo = gameInfo
+        this.tileMap = gameInfo.tileMap
+
+        // Add any preexisting camps as Encampment objects
+        for (tile in tileMap.values) {
+            if (tile.improvement == Constants.barbarianEncampment
+                && camps.none { it.position == tile.position }) {
+                val newCamp = Encampment()
+                newCamp.position = tile.position
+                newCamp.gameInfo = gameInfo
+                camps.add(newCamp)
+            }
+        }
+
+        for (camp in camps)
+            camp.gameInfo = gameInfo
+    }
+
+    fun updateEncampments() {
+        // Check if camps were destroyed
+        for (camp in camps.toList()) {
+            if (tileMap[camp.position].improvement != Constants.barbarianEncampment) {
+                camps.remove(camp)
+            }
+        }
+
+        // Possibly place a new encampment
+        placeBarbarianEncampment()
+
+        // Update all existing camps
+        for (camp in camps) {
+            camp.update()
+        }
+    }
+
+    /** Called when an encampment was attacked, will speed up time to next spawn */
+    fun campAttacked(position: Vector2) {
+        camps.firstOrNull { it.position == position }?.wasAttacked()
+    }
+
+    fun placeBarbarianEncampment() {
+        // Before we do the expensive stuff, do a roll to see if we will place a camp at all
+        if (gameInfo.turns > 1 && Random().nextBoolean())
+            return
+
+        // Barbarians will only spawn in places that no one can see
+        val allViewableTiles = gameInfo.civilizations.filterNot { it.isBarbarian() || it.isSpectator() }
+            .flatMap { it.viewableTiles }.toHashSet()
+        val fogTiles = tileMap.values.filter { it.isLand && it !in allViewableTiles }
+
+        val fogTilesPerCamp = (tileMap.values.size.toFloat().pow(0.4f)).toInt() // Approximately
+
+        // Check if we have more room
+        var campsToAdd = (fogTiles.size / fogTilesPerCamp) - camps.size
+
+        // First turn of the game add 1/3 of all possible camps
+        if (gameInfo.turns == 1) {
+            campsToAdd /= 3
+        } else if (campsToAdd > 0)
+            campsToAdd = 1
+
+        if (campsToAdd <= 0) return
+
+        // Camps can't spawn within 7 tiles of each other or within 4 tiles of major civ capitals
+        val tooCloseToCapitals = gameInfo.civilizations.filterNot { it.isBarbarian() || it.isSpectator() || it.cities.isEmpty() || it.isCityState() }
```

Comment:
Depending on the sizes of the tile collections it may be more efficient to not keep two overlapping sets of tiles to check for non-inclusion - keeping the capitals as iterable may suffice, then use .none { ... aerialDistanceTo ... }. Otherwise, when staying with sets, sets of Vector2 are more lightweight than those of TileInfo - and HashSets or -Maps of TileInfo are everywhere but theoretically unreliable as TileInfo doesn't do correct implementations of hashCode and equals - and building larger Sets has a smaller footprint if sequenced. May be academic. Thoughts triggered, goal of comment already reached.

Reply:
.none { ... aerialDistanceTo ... } was actually my first attempt, but this is significantly faster, likewise with mapping to Vector2. Of course this is on a PC and memory may be more of an issue on old phones, I'm not knowledgeable enough to tell. I'll let this stand unless you think it will actually be an issue for some users.

Reply:
No no I just like to push, but in the end I trust you to use your brain.
Measuring memory is - hard to do on the JVM. Easier to (guess the memory footprint of a single instance) * (guess cardinality) * (JVM collections overhead). With Vector2 ~8 Byte and TileInfo wild guess 100 Bytes - a large map can have 5000 - still less than 1 MB. But if you allocate and GC-abandon that regularly, markedly increased GC pauses may crop up much later...

Reply:
A way to test would be - console log a marker, deploy & run on a stone age AVD with dalvik and watch the GC messages on the logcat close to that marker, before/after. :grin: :skunk:

Reply:
I would expect the Set of TileInfos to just contain references to the actual TileInfos held by TileMap. Experimentally doing something silly to the tiles in the final viableTiles list has an effect on the game map.

Reply:
:facepalm: oops silly me. Of course those maps don't hold clones. So, just keep the equality issue in mind as decision argument. Why Maps of TileInfo work at all may be luck, may be some deterministic secret (probably no map ever sees both a tile and a clone of it), but in theory - I had that kind of failure in KeyPressDispatcher and took the easy route to make KeyCharAndCode properly obey equality (data).